src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +28,239 @@
 
-        do {
-            pet = new Pet();
+            String name = fields[PET_NAME_INDEX];
+            String dateString = fields[DATE_INDEX];
+            String typeIDString = fields[TYPE_INDEX];
+            String ownerIDString = fields[OWNER_ID_INDEX];
+            String commandType = fields[COMMAND_TYPE_INDEX].toLowerCase();
 
-            String field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
+            String errorMessage = validateCommand(fields, command);
+            if(!errorMessage.isEmpty()){
+                return createHttpError(errorMessage, HttpStatus.BAD_REQUEST);
             }
-            i++;
 
-            pet.setName(field);
+            Date date = parseDate(dateString);
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
+            int ownerID = Integer.parseInt(ownerIDString);
+            Owner owner;
 
             try {
-                pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(field));
-            } catch (ParseException e) {
-                HttpHeaders headers = new HttpHeaders();
-                headers.add("errors", "date " + field + " not valid");
-                return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
+                owner = clinicService.findOwnerById(ownerID);
+            }catch (ObjectRetrievalFailureException|EmptyResultDataAccessException e){
+                errorMessage = "In command: " + command + "\n";
+                errorMessage += "Owner is not found, given this ID: " + ownerID;
+                return createHttpError(errorMessage, HttpStatus.NOT_FOUND);
             }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
+            int typeID = Integer.parseInt(typeIDString);
+            PetType type;
+
+            try {
+                type = clinicService.findPetTypeById(typeID);
+            }catch (ObjectRetrievalFailureException|EmptyResultDataAccessException e){
+                errorMessage = "In command: " + command + "\n";
+                errorMessage += "Pet type is not found, given this ID: " + typeID;
+                return createHttpError(errorMessage, HttpStatus.NOT_FOUND);
             }
-            i++;
 
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
+            Pet pet = createPet(name, date, type);
+
+            // Beware of setting owner before deletion. This will result in no deletion occurring.
+            if(commandType.equals("add")){
+                pet.setOwner(owner);
```

Comment:
why don't we set the owner already in line 100?

Reply:
I thought `setOwner` updates the list of pets of the given owner, in this case the deletion wouldn't have happened. Also the delete command doesn't require the owner attribute.

Reply:
but you delete pet1 and not pet and your new equals methods those not consider the owners - so it should be able to set the owner as suggested by edmond?

Reply:
Yes, it should work. I just thought incorrectly that `setOwner` adds `pet` in the owner's pet list, but it doesn't.